### PR TITLE
Add editor setting to keep bottom panel state on play and stop game

### LINKED ISSUE #8507: Allow the debugger panel to stay closed on play unless there are runtime errors
### Describe the project you are working on

Any game

### Describe the problem or limitation you are having in your project

Every time I run the project, the debugger panel pops out, even if it doesn't have anything interesting to say. For example it opens when I only have a bunch of warnings that I don't care about:

https://github.com/godotengine/godot-proposals/assets/60579014/1c34ab19-eb00-419d-88d8-01c094b604a4

So every time I run the game I then have to manually close the bottom panel. This is especially distracting when I'm using the distraction-free mode

Also there's no way to close it with a hotkey, see #380

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

There could be a dropdown in editor settings to change that behavior

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The dropdown could be:
"`Open Debugger Panel`": "`Always On Play`/`On Errors Only`/`On Errors or Warnings`/`Never`"

The dropdown could be in the Debugger section of editor settings:

debugger

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't think so

### Is there a reason why this should be core and not an add-on in the asset library?

QOL editor improvement, too basic to install an addon for this

### LINKED ISSUE #9513: Add editor setting to keep bottom panel state on play and stop game
### Describe the project you are working on

A 3D air combat game

### Describe the problem or limitation you are having in your project

I want to keep the current bottom panels as it is when I start and exit my game. 

Each time I start my game from the editor, the Output Panel opens even if the bottom panels are closed. When I'm editing GDScript I like to collapse the bottom panels to have more space, but when I start/close the game and go back to the editor, the Output Panel is now opened.

There is already an option to Open the Output panel instead of the Debugger and an option to close it when exiting the game.
image

The problem with the current options is that I cannot just tell the editor to do nothing on start. If I use the option "Always close Output on Stop" I lose the current bottom panel when I exit the game.

My suggest is similar to #8507 but there is often errors thrown by the engine that are not important (especially when working in dev build).

Example of workflow #1:
1. Editing script...
image

2. Play... We can see the Output panel in the way...
image

3. Exit game, when I come back in the Editor... The bottom panel is in the way...
image

----------
Example of workflow #2:
1. Editing Animation...
image

2. Play... The animation panel is gone...
image

3. Exit game... I must click Animation again to go back editing the animation...
image

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The problem is backward compatibility with the current options. Currently, if the "Always Output on Play" option is Off, the debugger opens on start.
I suggest adding the option "Keep current bottom panel on play" with the default value false:
image

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

If this new option is On, the editor will not open or collapse bottom panels when starting or stopping a game.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't think it's possible.

### Is there a reason why this should be core and not an add-on in the asset library?

I'm not sure it can be done with a plugin because it's builtin in the editor code:
image

### PR DESCRIPTION
**Pull request by Hilderin**
This PR adds the editor settings run/output/keep_current_bottom_panel_on_play. The setting is false by default keep the old behavior.
It should close the proposal: https://github.com/godotengine/godot-proposals/issues/9513

image

The objective is to override the settings run/output/always_close_output_on_stop and run/output/keep_current_bottom_panel_on_play so the bottom panel stay the way it was when starting and stopping the game.

*Production edit:*
- Closes https://github.com/godotengine/godot-proposals/issues/8507
- Closes https://github.com/godotengine/godot-proposals/issues/9513

### CODE REVIEW

File: editor/editor_settings.cpp
Comment (akien-mga) [inline editor/editor_settings.cpp:1507]: Since this is compatibility code, we would usually make it conditional to this define:
```suggestion
#ifndef DISABLE_DEPRECATED
void EditorSettings::migrate_settings() {
	// Migrate the setting Always Open Output On Play that was True by default.
	// If false, then switch Action On Play to "Do Nothing" to keep same behavior.
	if (has_setting("run/output/always_open_output_on_play")) {
		if (!bool(get("run/output/always_open_output_on_play"))) {
			set("run/output/action_on_play", 1);
			erase("run/output/always_open_output_on_play");
		}
	}

	// Migrate the setting Always Close Output On Stop that was False by default.
	// If true, then switch Action On Stop to "Close Bottom Panel" to keep same behavior.
	if (has_setting("run/output/always_close_output_on_stop")) {
		if (bool(get("run/output/always_close_output_on_stop"))) {
			set("run/output/action_on_stop", 0);
			erase("run/output/always_close_output_on_stop");
		}
	}
}
#endif
```

Likewise in the header, and the place where it's called.

File: editor/editor_settings.cpp
Comment (akien-mga) [inline editor/editor_settings.cpp:822]: Since this is no longer specific to the Output panel, we should maybe rename the category.

Maybe `run/bottom_panel/` ?

File: editor/editor_settings.cpp
Comment (akien-mga) [reply editor/editor_settings.cpp:1507]: Actually though, I'm not sure we should bother migrating those settings:
- We've changed a lot of settings in the past and never migrated. Users for the most part seem to accommodate fine with having to retweak a few settings when upgrading.
- Keeping the old settings means that the configuration still works on older editor versions. This isn't as relevant following #90875 though in this particular context of making this change in 4.3 (but if we were to backport this to 4.2, it would be relevant).
- Since we add new options, this is an opportunity for users to reconsider their preferred option and maybe discover a behavior that will improve their experience.

**Comment by akien-mga** [comment]
I think this is a good approach :thumbsup:

Given the popularity of https://github.com/godotengine/godot-proposals/issues/8507 though, I wonder if we should consider making the default for "Action on Play" also to do nothing?

We could trial it in 4.3 beta, and see if the feedback we get is mostly that people miss having the Output dock automatically displayed, or actually like this new behavior.

One potential concern would be for new users to outright miss the fact that there's an Output panel where messages from their running game are displayed. There might be other ways to raise awareness about it, like adding a `(*)` in the title to show that there are unread messages or something? Ideas welcome.

File: editor/editor_settings.cpp
Comment (passivestar) [reply editor/editor_settings.cpp:822]: Was about to suggest this. The 2 old output-related ones can be put into an "Output" subsection (because you woudn't know what "Font Size" refers to otherwise)

**Comment by passivestar** [comment]
"Open Debugger" currently opens the output panel for me

**Comment by passivestar** [comment]
Also I just randomly found another PR that can be closed when this one is merged: https://github.com/godotengine/godot/pull/66503

**Comment by akien-mga** [comment]
> Also I just randomly found another PR that can be closed when this one is merged: #66503

And while at it, see also:
- #82277
- for https://github.com/godotengine/godot-proposals/issues/7848

**Comment by passivestar** [comment]
> One potential concern would be for new users to outright miss the fact that there's an Output panel where messages from their running game are displayed

@akien-mga Was it ever discussed to add an option to overlay output in the game itself like unreal does? I feel like godot might benefit from that a lot since it runs the game in a separate window and it can be inconvenient to see the output in a single-monitor setup

**Comment by Hilderin** [comment]
> "Open Debugger" currently opens the output panel for me

Effectively, I used the wrong value in `EditorNode::_project_run_started`
It will be fixed in my next commit.

File: editor/editor_settings.cpp
Comment (Hilderin) [reply editor/editor_settings.cpp:1507]: I removed the setting migration. Clearly it was professional deformation on my part, I developed management systems for the last 25 years and migration settings was always essential! It will be removed in my next commit.
I like your thinking!

File: editor/editor_settings.cpp
Comment (Hilderin) [reply editor/editor_settings.cpp:822]: I moved them in a new section "Bottom panel":
image

**Comment by Hilderin** [comment]
> Given the popularity of [godotengine/godot-proposals#8507](https://github.com/godotengine/godot-proposals/issues/8507) though, I wonder if we should consider making the default for "Action on Play" also to do nothing?

I changed the defaut value for "Action on Play" to "Do nothing". I'll wait for more feedback to see if we change it back to "Open Output".

**Comment by passivestar** [comment]
Opening debugger works now

I have this in Output though:

image<br>

If you make the debugger keep the currently opened tab instead of switching to stack trace this PR will also close https://github.com/godotengine/godot-proposals/issues/7848. But I don't know how easy it is to do, so it's your call

**Comment by Hilderin** [comment]
> I have this in Output though:

Hum... very weird... I don't have this behavior and I don't see how it is even possible, this setting has been removed from editor_settings.cpp!! Someone have a suggestion to fix this?!?!
image

**Comment by Hilderin** [comment]
> If you make the debugger keep the currently opened tab instead of switching to stack trace this PR will also close [godotengine/godot-proposals#7848](https://github.com/godotengine/godot-proposals/issues/7848). But I don't know how easy it is to do, so it's your call

I'll look into it.

**Comment by passivestar** [comment]
> Hum... very weird... I don't have this behavior and I don't see how it is even possible, this setting has been removed from editor_settings.cpp!! Someone have a suggestion to fix this?!?!

If I manually remove it from my `editor-settings-4.tres` it's gone, I don't know if it's something that will happen automatically when people update editor versions. iirc there are now also minor versions of editor settings (`editor-settings-4.3.tres`, etc), but I don't know how it migrates, if it's blindly copies it or actually checks individual settings. Somebody who knows will need to clarify

**Comment by passivestar** [comment]
> Ok, it was kinda too easy

Tested, works as expected

@akien-mga This PR now also closes both of https://github.com/godotengine/godot-proposals/issues/7848 and https://github.com/godotengine/godot-proposals/issues/3480

**Comment by Hilderin** [comment]
I placed "Do nothing" first in both settings, that way, the default option is the first and it will be easier to add new options in the future without having "Do nothing" in the middle.

Also, I found a problem with the "Close bottom panel" on stop, it was closing the bottom panel only if current panel was Output. It now always closes the bottom panel whatever panel you are on. This changes the behavior of the previous option "Always close on stop". I think it's ok and makes more sense??

**Comment by passivestar** [comment]
> Also, I found a problem with the "Close bottom panel" on stop, it was closing the bottom panel only if current panel was Output. It now always closes the bottom panel whatever panel you are on. This changes the behavior of the previous option "Always close on stop". I think it's ok and makes more sense??

This makes sense to me, I could imagine somebody for example messing with the debugger when the game is running and wanting it to close when the game stops

If we dive deep we'd probably find an even better design for all of this but tbh I think it's good enough at this point. This PR solves a bunch of actual reported problems, including mine

**Comment by KoBeWi** [comment]
> I don't know if it's something that will happen automatically when people update editor versions.

It doesn't. See https://github.com/godotengine/godot-proposals/issues/4444

File: editor/debugger/editor_debugger_node.cpp
Comment (KoBeWi) [inline editor/debugger/editor_debugger_node.cpp:273]: It would be better to replace magic numbers with some enum (defined in one place), even if EditorSettings have to repeat it.

File: editor/editor_node.cpp
Comment (KoBeWi) [inline editor/editor_node.cpp:4394]: I know this was here originally, but I wonder why it has to be done twice? (the other one being in debugger)
This one doesn't even handle the Open Debugger case; maybe it can be removed.

Comment (Hilderin) [reply editor/editor_node.cpp:4394]: Hehe, I asked myself the same thing! I don't known, I figured at the time that there was a reason. I'll remove it.

Comment (KoBeWi) [reply editor/editor_node.cpp:4394]: Or you could remove the other one, so that start and stop are handled in the same file. Not sure if the debugger is available here though.

**Comment by Hilderin** [comment]
This last commit adds 2 enums and move the code from editor_debugger_node to editor_node which effectively is much cleaner.
thanks @KoBeWi

File: editor/editor_node.h
Comment (KoBeWi) [inline editor/editor_node.h:150]: Due to how enums work in C++ you should either prefix them with name (e.g. `STOP_DO_NOTHING`), or I think using `enum class` would fix it too.

Comment (Hilderin) [reply editor/editor_node.h:150]: Yeah, I'll adjust, I'm not very experienced with different versions of C++, it was working for me, I'll adjust.

**Review by KoBeWi — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected.

I think this new default makes sense, but I'm worried about users potentially not noticing where their `print()` output goes (particularly for beginners). I wonder if we should make it so the Output panel is opened if you currently don't have any panel open.

Perhaps this could also be alleviated by adding an "unread messages count" to the Output panel button's text, like Debugger already gets when warnings/errors are printed. I'd prefer this solution personally, but it should probably be implemented in a separate PR.

**Comment by Hilderin** [comment]
I understand your concern and Ilike your idea of unread count like the debugger. I think when you start Godot for the first time the default layout has the Output panel open. In fact, I remember as a new user searching how to close this panel at the bottom. My point is if the Ouput panel is there by default, the user should see the output by default.

**Comment by Hilderin** [comment]
I'm thinking about what @Calinou said and I don't see a lot of possibilities. The easy solution is to change the default value for `Action on play` to be `Open output`, which will be equivalent to Godot 4.2.

I'm not sure about the suggestion "I wonder if we should make it so the Output panel is opened if you currently don't have any panel open." because I personally don't use the builtin print. I have a custom print that prints output directly in the game (similar to Unreal) and Godot prints thinks that I don't need.

**Comment by KoBeWi** [comment]
> I'm not sure about the suggestion "I wonder if we should make it so the Output panel is opened if you currently don't have any panel open." because I personally don't use the builtin print.

It could be one of the available options, but IMO it's too specific.

**Comment by akien-mga** [comment]
Given the above discussion, I think we can still merge this now to include in 4.3 (despite feature freeze - this is a relatively safe change and with a lot of demand), but given the timeline I would opt for using defaults that match the pre-existing behavior for now.

We won't have much time during the short beta period for collecting user feedback about potential new defaults, so I think it's safer to provide the extra features while keeping the same defaults for 4.3. And we can try out changing the defaults for 4.4 early on the dev cycle, so we have several months to gather and address user feedback.

**Comment by Hilderin** [comment]
Wonderful! I changed the default value for `Action on play` to be `Open output`, which should match the pre-existing behavior.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
